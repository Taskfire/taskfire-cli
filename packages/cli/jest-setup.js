import dotenv from 'dotenv'
import path from 'path'

/* istanbul ignore next */
if (typeof process === 'object' && !global.unhandledRejection) {
  global.unhandledRejection = true
  process.on('unhandledRejection', (error) => {
    console.error(error)
  })
}

if (process.env.NODE_ENV === 'test') {
  dotenv.config({
    path: path.resolve(__dirname, './.env.test'),
  })
}
