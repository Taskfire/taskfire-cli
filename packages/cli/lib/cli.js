#!/usr/bin/env node
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _yargs = require('yargs');

var _yargs2 = _interopRequireDefault(_yargs);

var _raven = require('raven');

var _raven2 = _interopRequireDefault(_raven);

var _flows = require('./cmds/flows');

var _flows2 = _interopRequireDefault(_flows);

var _projects = require('./cmds/projects');

var _projects2 = _interopRequireDefault(_projects);

var _login = require('./cmds/login');

var _login2 = _interopRequireDefault(_login);

var _signup = require('./cmds/signup');

var _signup2 = _interopRequireDefault(_signup);

var _runs = require('./cmds/runs');

var _runs2 = _interopRequireDefault(_runs);

var _run = require('./cmds/run');

var _run2 = _interopRequireDefault(_run);

var _logs = require('./cmds/logs');

var _logs2 = _interopRequireDefault(_logs);

var _images = require('./cmds/images');

var _images2 = _interopRequireDefault(_images);

var _variables = require('./cmds/variables');

var _variables2 = _interopRequireDefault(_variables);

var _deploy = require('./cmds/deploy');

var _deploy2 = _interopRequireDefault(_deploy);

var _config = require('./cmds/config');

var _config2 = _interopRequireDefault(_config);

var _init = require('./cmds/init');

var _init2 = _interopRequireDefault(_init);

var _output = require('./helpers/output');

var _output2 = _interopRequireDefault(_output);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const cli = _yargs2.default.usage('Usage: taskfire <command> [options]').scriptName('taskfire').command(_flows2.default).command(_projects2.default).command(_runs2.default).command(_run2.default).command(_logs2.default).command(_deploy2.default).command(_images2.default).command(_variables2.default).command(_config2.default).command(_login2.default).command(_signup2.default).command(_init2.default).option('token', {
  alias: 't',
  describe: 'Authentication token'
}).option('project', {
  alias: 'p',
  describe: 'Project to target'
}).option('silent', {
  alias: 's',
  describe: 'Silent mode (no stdout)'
}).strict().demandCommand(1).fail((msg, err) => {
  console.log('ERROR', msg, err);
  if (err) {
    // Need to make sure we filter out known errors
    _raven2.default.captureException(err);
  }
  _output2.default.space();
  if (msg || err && err.message) {
    _output2.default.accent(msg || err && err.message);
    _output2.default.space();
  }
  _yargs2.default.showHelp();
}).recommendCommands().help('h');
// .help('help')
// .parse()

exports.default = cli;