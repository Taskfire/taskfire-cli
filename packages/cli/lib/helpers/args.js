'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProjectName = exports.getCwd = undefined;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _config = require('./config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getCwd = exports.getCwd = args => {
  return _path2.default.resolve(process.cwd(), args.dir || '');
};

const getProjectName = exports.getProjectName = async args => {
  // 1. Use the flag
  if (args.p || args.project) return args.p || args.project;

  // 2. Use the directory .taskfirerc
  const cwd = getCwd(args);
  const taskfirercPath = _path2.default.resolve(cwd, '.taskfirerc');
  if (_fsExtra2.default.existsSync(taskfirercPath)) {
    const taskfirerc = await _fsExtra2.default.readJson(_path2.default.resolve(cwd, '.taskfirerc'), { throws: false });
    if (taskfirerc && taskfirerc.project) return taskfirerc.project;
  }

  // 3. Use the default for the user
  return (0, _config.getBasicConfig)(args, 'defaultProject');
};