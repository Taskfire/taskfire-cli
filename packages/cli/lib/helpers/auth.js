'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _errors = require('./errors');

var _errors2 = _interopRequireDefault(_errors);

var _config = require('./args/config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const NO_TOKEN_MSG = `
  ${_chalk2.default.red('No authentication token')}

  Use ${_chalk2.default.cyanBright('$ taskfire login')} to login
  Or ${_chalk2.default.cyanBright('$ taskfire signup')} to create a new account
`;

exports.default = async function getAuth(args, requireAuth = true) {
  let token = args.t || args.token;

  if (!token) {
    // TODO: Get token from the cache
    const auth = await (0, _config.getAuthConfig)(args);
    if (auth) {
      // eslint-disable-next-line
      token = auth.credentials[auth.defaultAuth].token;
    }
  }

  if (!token && requireAuth) {
    (0, _errors2.default)(NO_TOKEN_MSG);
  }

  return token;
};