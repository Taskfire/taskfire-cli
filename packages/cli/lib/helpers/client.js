'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _cliSelect = require('cli-select');

var _cliSelect2 = _interopRequireDefault(_cliSelect);

var _taskfire = require('taskfire');

var _taskfire2 = _interopRequireDefault(_taskfire);

var _auth = require('./auth');

var _auth2 = _interopRequireDefault(_auth);

var _output = require('./output');

var _output2 = _interopRequireDefault(_output);

var _config = require('./config');

var _args = require('./args');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import taskfire from '../../../taskfire-nodejs/lib/index'
exports.default = async function createClient(args, requireAuth, requireSite = false) {
  const auth = await (0, _auth2.default)(args, requireAuth, requireSite);
  const projectName = await (0, _args.getProjectName)(args);

  const client = (0, _taskfire2.default)(auth.token, {
    project: projectName,
    url: process.env.NODE_ENV === 'development' ? 'http://localhost:4000' : 'https://api.taskfire.io'
    // debug: true,
  });

  // If we need a project/site then we should
  // ask the user for it (as no default is present)
  if (requireSite && !projectName) {
    const projects = await client.projects.list();

    if (!projects.length) {
      _output2.default.error('You must create a project before deploying your code');
    }

    // Ask the user to select a project
    const selectedIndex = await (0, _cliSelect2.default)({
      values: projects.map(({ name }) => name)
    }).catch(() => {
      _output2.default.error('Project ID required');
    });

    const selectedProject = projects[selectedIndex.id];

    // Set it as the default for the future
    await (0, _config.setBasicConfig)(args, 'defaultProject', selectedProject.name);

    (0, _output2.default)(`Setting ${selectedProject.name} as default config`);

    // Update the client with the selected projectId
    client.options.project = selectedProject.name;
  }

  return client;
};