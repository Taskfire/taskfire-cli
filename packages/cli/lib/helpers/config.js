'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getConfigPath;
exports.getAuthConfig = getAuthConfig;
exports.addAuthConfig = addAuthConfig;
exports.setBasicConfig = setBasicConfig;
exports.getBasicConfig = getBasicConfig;

var _os = require('os');

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _loadJsonFile = require('load-json-file');

var _loadJsonFile2 = _interopRequireDefault(_loadJsonFile);

var _writeJsonFile = require('write-json-file');

var _writeJsonFile2 = _interopRequireDefault(_writeJsonFile);

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _set = require('lodash/set');

var _set2 = _interopRequireDefault(_set);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const dirName = process.env.NODE_ENV === 'development' ? '.taskfire.dev' : '.taskfire';

const defaultAuth = {
  _: 'This is your Taskfire credentials file. DON\'T SHARE!',
  credentials: {}
};

const getAuthConfigPath = args => _path2.default.resolve(getConfigPath(args), './auth.json');

const getBasicConfigPath = args => _path2.default.resolve(getConfigPath(args), './config.json');

function getConfigPath(args) {
  const { configPath } = args;
  if (configPath) {
    return _path2.default.resolve(configPath);
  }
  return _path2.default.resolve((0, _os.homedir)(), `./${dirName}'`);
}

async function getAuthConfig(args) {
  const authPath = getAuthConfigPath(args);
  return _fsExtra2.default.existsSync(authPath) ? (0, _loadJsonFile2.default)(authPath) : null;
}

async function addAuthConfig(args, username, token) {
  const authPath = getAuthConfigPath(args);

  let auth = await getAuthConfig(args);

  if (!auth) {
    auth = defaultAuth;
  }

  // Update the auth config
  auth.credentials[username] = {
    type: 'jwt',
    token
  };
  auth.defaultAuth = username;

  await setBasicConfig('defaultProject', null);

  // Save the file
  return (0, _writeJsonFile2.default)(authPath, auth);
}

async function setBasicConfig(args, key, value) {
  const configPath = getBasicConfigPath(args);
  let json = {};
  if (_fsExtra2.default.existsSync(configPath)) {
    json = await _fsExtra2.default.readJson(configPath);
  }
  (0, _set2.default)(json, key, value);
  await _fsExtra2.default.outputJson(configPath, json, {
    spaces: 2
  });
  return value;
}

async function getBasicConfig(args, key) {
  const configPath = getBasicConfigPath(args);
  if (!_fsExtra2.default.existsSync(configPath)) return null;
  const json = await _fsExtra2.default.readJson(configPath);
  return (0, _get2.default)(json, key);
}