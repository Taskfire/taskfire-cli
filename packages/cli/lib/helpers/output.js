'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _forEach = require('lodash/forEach');

var _forEach2 = _interopRequireDefault(_forEach);

var _pad = require('pad');

var _pad2 = _interopRequireDefault(_pad);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function output(msg, args = {}) {
  if (args.s || args.silent) return;
  if (process.env.NODE_ENV !== 'test') console.log(msg);
}

output.block = function outputBlock(msg, args) {
  output(['', msg, ''].join('\n'), args);
};

output.error = function outputError(msg, exit = true) {
  output(_chalk2.default.red(msg));
  if (exit) process.exit(1);
};

output.success = function outputSuccess(msg, args) {
  output(_chalk2.default.green(msg), args);
};

output.info = function outputInfo(msg, args) {
  output(`${_chalk2.default.cyan('[INFO]')} ${msg}'`, args);
};

output.accent = function outputAccent(msg, args) {
  output(`${_chalk2.default.grey('>')} ${msg}`, args);
};

output.space = function outputSpace(args) {
  output('', args);
};

output.record = function outputRecord(record, args) {
  let length = 0;

  // Get the maximum length
  (0, _forEach2.default)(record, (val, key) => {
    length = Math.max(key.length + 2, length);
  });

  (0, _forEach2.default)(record, (val, key) => {
    const name = `${key}:`;
    output(`${(0, _pad2.default)(name, length)}${val}`, args);
  });
};

exports.default = output;