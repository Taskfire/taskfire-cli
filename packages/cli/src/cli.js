#!/usr/bin/env node

import yargs from 'yargs'
import flows from './cmds/flows'
import projects from './cmds/projects'
import login from './cmds/login'
import signup from './cmds/signup'
import runs from './cmds/runs'
import run from './cmds/run'
import logs from './cmds/logs'
import images from './cmds/images'
import variables from './cmds/variables'
import deploy from './cmds/deploy'
import config from './cmds/config'
import init from './cmds/init'
import output from './helpers/output'

const cli = yargs
  .usage('Usage: taskfire <command> [options]')
  .scriptName('taskfire')
  .command(flows)
  .command(projects)
  .command(runs)
  .command(run)
  .command(logs)
  .command(deploy)
  .command(images)
  .command(variables)
  .command(config)
  .command(login)
  .command(signup)
  .command(init)
  .option('token', {
    alias: 't',
    describe: 'Authentication token',
  })
  .option('project', {
    alias: 'p',
    describe: 'Project to target',
  })
  .option('silent', {
    alias: 's',
    describe: 'Silent mode (no stdout)',
  })
  .strict()
  .demandCommand(1)
  .fail((msg, err) => {
    console.log(err)
    output.space()
    if (msg || (err && err.message)) {
      output.accent(msg || (err && err.message))
      output.space()
    }
    yargs.showHelp()
  })
  .recommendCommands()
  .help('h')
  // .help('help')
  // .parse()

export default cli
