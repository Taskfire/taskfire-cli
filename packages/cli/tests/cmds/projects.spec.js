import cli from '../../src/cli'
import { checkOutput } from '../helpers/utils'

jest.mock('request-promise-native')

describe('cmds/projects.spec', () => {
  test('projects help command runs', async () => {
    const { exit, errors } = await checkOutput(() =>
      cli.parse(['projects', '-h']))

    expect(exit).toBe(true)
    expect(errors).toEqual([])
  })

  test('projects help command matches snapshot', async () => {
    const { logs } = await checkOutput(() =>
      cli.parse(['projects', '-h']))

    expect(logs).toMatchSnapshot()
  })

  // test.only('projects list command with no projects', async () => {
  //   taskfire.projects = {
  //     list: jest.fn().mockResolvedValue([]),
  //   }
  //
  //   const { exit, errors } = await checkOutput(() =>
  //     cli.parse(['projects', '--token', 't123', '--project', 'p123']))
  //
  //   //
  //   // expect(exit).toBe(true)
  //   // expect(errors).toEqual([])
  // })
})
