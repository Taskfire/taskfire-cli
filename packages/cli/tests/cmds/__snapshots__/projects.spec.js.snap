// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cmds/projects.spec projects help command matches snapshot 1`] = `
Array [
  "taskfire projects

Manage projects

Commands:
  taskfire projects list           List projects
  taskfire projects create [name]  Create a new project

Options:
  --version      Show version number                                   [boolean]
  --token, -t    Authentication token
  --project, -p  Project to target
  --silent, -s   Silent mode (no stdout)
  -h             Show help                                             [boolean]",
]
`;
